File: src/Layer.java
//-----[IMPORTS]-----\\

//-----[CLASS]-----\\
/**
 * Layer
 * Class
 * 
 * A layer of nodes in the neural network.
 * 
 * @author Joshua Savoie
 */
public class Layer {

    //-----[CONSTANTS]-----\\

    // Used for weight regularization and generation
    private static final double UPPER_THRESHOLD = 1.0;
    private static final double LOWER_THRESHOLD = -1.0;

    //-----[VARIABLES]-----\\

    // Text file to store data to
    private String m_manifestFile;

    // The number of nodes
    private final int m_layerSize;
    // The number of inputs
    private final int m_numInputs;

    // Biases for each node
    private double[] m_biases;
    // Weights for each node to each input
    private double[][] m_weights;

    // The change for each bias to descend the gradient slope
    private double[] m_biasesGradient;
    // The change for each weight to each input to descend the gradient slope
    private double[][] m_weightsGradient;

    // The most recent outputs of the layer without applying the activation function
    private double[] m_zVector;
    // The most recent outputs with the activation function applied
    private double[] m_outputVector;

    //-----[CONSTRUCTORS]-----\\

    /**
     * Layer
     * Constructor
     * 
     * Creates a neural network layer.
     * 
     * @param manifestFile - Text file to store to.
     * @param layerSize - Size of the layer.
     * @param numInputs - The number of inputs going into the layer.
     */
    public Layer(String manifestFile, int layerSize, int numInputs) {

        m_manifestFile = manifestFile;

        m_layerSize = layerSize;
        m_numInputs = numInputs;

        m_biases = new double[m_layerSize];
        m_weights = new double[m_layerSize][m_numInputs];

        m_biasesGradient = new double[m_layerSize];
        m_weightsGradient = new double[m_layerSize][m_numInputs];

        // Sets all of the initial values for the weights and biases
        for(int n = 0; n < m_layerSize; n++) {

            m_biases[n] = 0;
            m_biasesGradient[n] = 0;

            for(int w = 0; w < m_numInputs; w++) {
                // Generates a weight value between the upper and lower thresholds.
                m_weights[n][w] = Math.random() * (UPPER_THRESHOLD - LOWER_THRESHOLD) + LOWER_THRESHOLD;
                m_weightsGradient[n][w] = 0;
            }
        }
    }

    /**
     * Layer
     * Constructor
     * 
     * Constructor only meant for copying a layer.
     * 
     * @param manifestFile
     * @param layerSize
     * @param numInputs
     * @param biases
     * @param weights
     * @param biasesGradient
     * @param weightsGradient
     * @param zVector
     * @param outputVector
     */
    private Layer(
        String manifestFile, int layerSize, int numInputs, 
        double[] biases, double[][] weights, 
        double[] biasesGradient, double[][] weightsGradient, 
        double[] zVector, double[] outputVector
    ) {
        m_manifestFile = manifestFile;
        m_layerSize = layerSize;
        m_numInputs = numInputs;
        m_biases = biases;
        m_weights = weights;
        m_biasesGradient = biasesGradient;
        m_weightsGradient = weightsGradient;
        m_zVector = zVector;
        m_outputVector = outputVector;
    }

    //-----[METHODS]-----\\

    /**
     * calculate()
     * 
     * Given some inputs, calculate the outputs for each node in this layer.
     * 
     * @param inputs - The input values into the layer.
     * @return
     */
    public double[] calculate(double[] inputs) {

        assert m_numInputs == inputs.length: " the input size of the layer and the number of inputs do not match.";

        // Outputs with activation function
        double[] outputs = new double[m_layerSize];
        // Outputs without activation function
        m_zVector = new double[m_layerSize];

        // summation(w * i + b)
        for(int n = 0; n < m_layerSize; n++) {
            
            double out = m_biases[n];

            for(int i = 0; i < m_numInputs; i++) {
                out += (m_weights[n][i] * inputs[i]);
            }

            m_zVector[n] = out;

            // Applies the activation function
            outputs[n] = Functions.bipolarSigmoid(out);

        }

        // Stores a copy of the output vector with the activation function applied
        m_outputVector = outputs.clone();

        return outputs;

    }

    /**
     * updateGradient()
     * 
     * Updates the gradient slope according to this layer's (L) error and the previous layer's (L-1) activations.
     * 
     * @param errors - This layer's (L) error.
     * @param previousActivations - The previous layer's (L-1) activations.
     */
    public void updateGradient(double[] errors, double[] previousActivations) {

        assert errors.length == m_layerSize: " the number of errors does not match the number of nodes.";
        assert previousActivations.length == m_numInputs: " the number of activations provided does not match the number of inputs to this layer.";
        
        for(int n = 0; n < m_layerSize; n++) {

            // dC/db = error
            // Change of cost in terms of bias equals the node's error
            m_biasesGradient[n] += (errors[n]);

            // dC/dw = activation * error
            // Change of cost in terms of weight equals the previous layer's activation multiplied by the node's error
            for(int w = 0; w < m_numInputs; w++) {
                m_weightsGradient[n][w] += (previousActivations[w] * errors[n]);
            }
        }

    }

    /**
     * applyGradient()
     * 
     * Applies the gradient to the layer.
     * 
     * @param miniBatchSize - The size of the training mini-batch.
     * @param learnRate - The rate to descend the gradient slope.
     */
    public void applyGardient(int miniBatchSize, double learnRate) {

        for(int n = 0; n < m_layerSize; n++) {

            // Change
            double d = 0;

            // Averages the gradient against the size of the mini-batch size.
            // Equivalent to storing all of the gradients from each training data point but saves a lot of space.
            d = m_biasesGradient[n] / miniBatchSize * learnRate;

            // Bias regularization after exceeding a threshold
            if(
                (m_biases[n] > UPPER_THRESHOLD && d > 0) ||
                (m_biases[n] < LOWER_THRESHOLD && d < 0)
            ) {
                m_biases[n] -= (d * (1.0 / Math.abs(m_biases[n])));
            } else {
                m_biases[n] -= d;
            }

            m_biasesGradient[n] = 0;

            for(int w = 0; w < m_numInputs; w++) {

                d = m_weightsGradient[n][w] / miniBatchSize * learnRate;

                // Weight regularization after exceeding a threshold
                if(
                    (m_weights[n][w] > UPPER_THRESHOLD && d > 0) ||
                    (m_weights[n][w] < LOWER_THRESHOLD && d < 0)
                ) {
                    m_weights[n][w] -= (d * (1.0 / Math.abs(m_weights[n][w])));
                } else {
                    m_weights[n][w] -= d;
                }

                m_weightsGradient[n][w] = 0;

            }
        }
    }

    /**
     * getZVector()
     * 
     * Gets the outputs of the layer without the activation function applied.
     * 
     * @return
     */
    public double[] getZVector() {
        return m_zVector.clone();
    }

    /**
     * getOutputVector()
     * 
     * Gets the outputs of the layeer with the activation function applied.
     * 
     * @return
     */
    public double[] getOutputVector() {
        return m_outputVector.clone();
    }

    /**
     * getWeights()
     * 
     * Gets the weights of the layer.
     * 
     * @return
     */
    public double[][] getWeights() {
        return m_weights.clone();
    }

    /**
     * getLayerSize()
     * 
     * Gets the size of the layer.
     * 
     * @return
     */
    public int getLayerSize() {
        return m_layerSize;
    }

    /**
     * copy()
     * 
     * Returns a copy of the scoped layer.
     * 
     * @return
     */
    public Layer clone() {
        return new Layer(
            new String(m_manifestFile), m_layerSize, m_numInputs, 
            m_biases.clone(), m_weights.clone(), 
            m_biasesGradient.clone(), m_weightsGradient.clone(), 
            m_zVector.clone(), m_outputVector.clone()
        );
    }

}

File: src/Layer.py
import math
import random
import threading
from pandas import DataFrame
from enum import Enum
from math import isinf, isnan
from typing import Callable
from concurrent.futures import ThreadPoolExecutor

class _Prefix(Enum):
    NODE = "n"
    BIAS = "b"
    WEIGHT = "w"

class Layer:

    #-------------[ CONSTRUCTORS ]-------------#
    
    def __init__(
        self, size: int, inputSize: int, 
        biases: list[float]=None, weights: list[list[float]]=None, 
        lower: float=0, upper: float=1
    ) -> object:
        """
        **Parameters**
        - *size: int*
        > Size of the layer. 
        > Determines the number of nodes.
        - *inputSize: int*
        > Number of inputs feeding into this layer. 
        > Determines the number of weights per node.
        - *biases: list[float]=None*
        > Optional.
        > Biases to build this layer from.
        > Length should match the size parameter.
        - *weights: list[float]=None*
        > Optional.
        > Weights to build this layer from.
        > Length should match the size parameter, and the width should match the inputSize parameter.
        """
                
        if(biases is None): self.biases = [] 
        else: self.biases = biases
        
        if(weights is None): self.weights = []
        else: self.weights = weights

        self.biasesGradient = []
        self.weightsGradient = []
    
        for n in range(size):
            
            if(biases is None): self.biases.append(0.0)
            if(weights is None): self.weights.append([])

            self.biasesGradient.append(0.0)
            self.weightsGradient.append([])
            
            for w in range(inputSize):
                
                if(weights is None): self.weights[n].append(random.uniform(lower, upper))

                self.weightsGradient[n].append(0.0)

    @classmethod
    def fromDataFrame(cls, df: DataFrame) -> object:
        """
        Nodes should be expressed as columns in the DataFrame.
        """
        
        biases = []
        weights = []
        for col in df:
            for i, v in df[col].items():
                if(isinf(float(v)) or isnan(float(v))):
                    raise ValueError(f"Value {v} is either infinite or nan.")
                match str(i)[0]:
                    case _Prefix.BIAS.value:
                        biases.append(float(v))
                    case _Prefix.WEIGHT.value:
                        weights.append(float(v))
                    case _:
                        raise KeyError(f"Index {i} does not match any of the enumerated keys: {[e.value for e in _Prefix]}")

        return cls(len(biases), len(weights[0]), biases, weights)
    
    def copy(self) -> object:
        nodes, weights = self.shape
        return Layer(
            size=nodes, inputSize=weights, 
            biases=self.biases, weights=self.weights
        )
            
    #-------------[ OBJECT DESCRIPTORS ]-------------#

    def __len__(self) -> int:
        return len(self.biases)
    
    @property
    def shape(self) -> tuple[int, int]:
        return (len(self.biases), len(self.weights[0]))
    
    @property
    def size(self) -> int:
        return len(self.biases) * len(self.weights[0]) + len(self.biases)
    
    def toDataFrame(self) -> DataFrame:

        data = {}
        indices = []
        nodes, weights = self.shape

        for j in range(nodes):

            if(j == 0):
                indices.append(f"{_Prefix.BIAS.value}")
            data[f"{_Prefix.NODE.value}{j}"] = [self.biases[j]]

            for k in range(weights):

                if(j == 0):
                    indices.append(f"{_Prefix.WEIGHT.value}{k}")
                data[f"{_Prefix.NODE.value}{j}"].append(self.weights[j][k])

        return DataFrame(data=data, index=indices)
    
    #-------------[ CLASS FUNCTIONS ]-------------#
                
    def calculate(self, inputs: list, activationFunction: Callable[[float, bool], float], adjustSize: bool=False):
        
        if(adjustSize):
            if(len(self) < len(inputs)):
                del self.biases[-1]
                del self.biasesGradient[-1]
                for n in range(len(self.weights)):
                    del self.weights[n][-1]
                    del self.weightsGradient[n][-1]
                    
            elif(len(self) > len(inputs)):
                self.biases.append(0.0)
                self.biasesGradient.append(0.0)
                for n in range(len(self.weights)):
                    self.weights[n].append(0.0)
                    self.weightsGradient[n].append(0.0)

        assert self.shape[1] == len(inputs)
        
        self.rawOutputs = []
        self.activatedOutputs = []

        nodes, weightsPerNode = self.shape
        for j in range(nodes):
            out = self.biases[j]
            for k in range(weightsPerNode):
                out += (self.weights[j][k] * inputs[k])
            
            self.rawOutputs.append(out) # Known as "z of L" in literature
            self.activatedOutputs.append(activationFunction(out, False)) # Known as "a of L" in literature
        
        return self.activatedOutputs.copy()
    
    def updateGradient(self, activationFunction: Callable[[float, bool], float], errors: list[float], backwardOutputs: list[float]) -> list[float]:
        """
        **Parameters**
        - *activationFunction: Callable*
        > The activation function to update the gradient with.
        - *errors: list[float]*
        > The errors calculated by the forward layer L+1 where L is this layer. 
        > For the output layer of a network, pass in the output errors resulting from the cost function.
        - *backwardOutputs: list[float]*
        > Activated outputs of the backward layer L-1.

        **Returns**
        - *backwardErrors: list[float]*
        > The backward errors for layer L-1. The hadamard product still needs to be performed on this array.
        > Used for backpropagation.
        """

        backwardErrors = []
        
        nodes, weights = self.shape
        for j in range(nodes):

            errors[j] = errors[j] * activationFunction(self.rawOutputs[j], True)
                # Performing the hadamard product node by node

            self.biasesGradient[j] += errors[j]

            for k in range(weights):
                # Weight from node j in layer L to node k in layer L-1 ==> w from k to j

                if(len(backwardErrors) < k + 1):
                    backwardErrors.append(0.0)
                backwardErrors[k] += self.weights[j][k] * errors[j]

                self.weightsGradient[j][k] += backwardOutputs[k] * errors[j]
                    # Don't need to multiply by the derived activation output of raw output z of node j since
                    # the hadamard product for j was computed earlier.

        return backwardErrors
    
    def applyGradient(self, batchSize: int, learnRate: float=1) -> None:
        
        nodes, weights = self.shape
        for j in range(nodes):
            
            self.biases[j] -= (self.biasesGradient[j] / batchSize * learnRate)
            self.biasesGradient[j] = 0.0
            
            for k in range(weights):
                
                self.weights[j][k] -= (self.weightsGradient[j][k] / batchSize * learnRate)
                self.weightsGradient[j][k] = 0.0
        